### ISSUE TITLE
[glsl-in/spv-in] Add support for BGRA8 storage textures somehow

### ISSUE THREAD

**Issue by ifiddynine** [gfx-rs/wgpu#4934]
**Description**
We use the swapchain texture views in a compute shader, which is compiled from GLSL -> SPIRV -> WGSL. We ran into an issue where at compute pipeline binding time the WGSL shader has to explicitly match the BGRA8 format, which actually isn't supported as an image layout in neither SPIRV or GLSL.

Caused by:
    In wgpuDeviceCreateComputePipeline
    Error matching shader requirements against the pipeline
    Shader global ResourceBinding { group: 1, binding: 3 } is not available in the layout pipeline layout
    Texture class Storage { format: Bgra8Unorm, access: StorageAccess(STORE) } doesn't match the shader Storage { format: Rgba8Unorm, access: StorageAccess(STORE) }

**Repro steps**
1. Create a compute shader that contains the following: `@group(0) @binding(0) var outputImage : texture_storage_2d<rgba8unorm, write>;`
2. Create a compute pipeline using `wgpuDeviceCreateComputePipeline` which uses the above compute shader and specifies the descriptor layout with format: Bgra8Unorm
4. Notice the above failure

**Expected vs observed behavior**
Since GLSL or SPIRV don't support specifying BGRA8 as an image format I'd argue that the validation error should not be emitted in this case, just like Metal and Vulkan treat them as interchangeable.  Otherwise supporting WGSL in a pipeline with shader tools would require non-standard shader compilation tools to be compatible and emit bgra8.

**Comment by jimblandy** [comment]
CC: @teoxoy 

Marking this as a Naga bug, but maybe it's wgpu?

**Comment by cwfitzgerald** [comment]
This seems like it's a problem of wgsl having more information than glsl and spirv less than a bug per-se.

An rgba8 format shouldn't be compatible with bgra8, as they are just not compatible. However, we should allow glsl/spirv-in to use the unspecified format specifier for storage images. This is how spirv-out is generating it, but our own spirv-in wouldn't properly handle it.

**Comment by teoxoy** [comment]
GLSL doesn't have a format qualifer for `bgra8`. The spec mentiones:

> When a texture is bound to an image unit, the format parameter for the image
unit need not exactly match the texture internal format as long as the formats are
considered compatible. A pair of formats is considered to match in size if the
corresponding entries in the Size column of table 8.27 are identical. A pair of formats
is considered to match by class if the corresponding entries in the Class column
of table 8.27 are identical. For textures allocated by the GL, an image unit format
is compatible with a texture internal format if they match by size. For textures
allocated outside the GL, format compatibility is determined by matching by size
or by class, in an implementation-dependent manner.

but Table 8.27 doesn't list any `bgra` formats.

_from https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf_

So from the OpenGL spec perspective, you can't use `bgra` formats for storage textures.

---

SPIR-V also doesn't have `bgra` image formats and the Vulkan spec also doesn't make bgra textures compatible with SPIR-V's `rgba` image formats.
The only way to use a `bgra` storage texture is via SPIR-V's `Unknown` image format.

---

The `GL_KHR_vulkan_glsl` extension also doesn't add any new format qualifers.

There doesn't seem to be anything we can do right now to support `bgra8` in our GLSL or SPIR-V front-ends.

I think we could omit the validation if we see SPIR-V's `Unknown` image format but you'd still need a GLSL compiler that implements a non-standard `bgra8` qualifier that gets translated to the `Unknown` image format. We could add the non-standard `bgra8` qualifier to our GLSL front-end but it sounds like you are using a different compiler and passing SPIR-V to wgpu instead.

**Comment by ifiddynine** [comment]
Yup we don't use any wglsl / wgpu specific GLSL compilers. Our pipeline is:
1. glslang / glslangValidator, glsl -> SPIR-V
2. spirvopt
3. tint -> wgsl (or SPIRV-Cross etc for Metal and other graphics APIs when running natively)

What's interesting is that Metal and Vulkan run fine natively when using BRGA8 (which often is the only swapchain format), I'm guessing they resolve this at a later point based on the descriptor layout and / or descriptor information we pass in. Haven't been able to check Dawn's behavior as their support for storage textures is behind what wgpu does.

Ideally for game engines we'd be able to plug into existing pipelines in the middle instead of replacing the whole flow for each platform & graphics API - I didn't know about the Unknown's behavior in SPIR-V though, will see if I can get our code to generate that within the existing pipeline and let you guys know who it works.

Thanks for the info!
